### DISCUSSION TITLE
Syntax for filtering documents

### DISCUSSION THREAD

**Discussion by kodumbeats** [appwrite/appwrite#1059] [Ideas]
# Client Filters for Documents 

We're in the planning phases of [refactoring our database](https://github.com/appwrite/rfc/pull/21), and _we need your help_. We aim to create the **best developer experience** while we build Appwrite, and we want document filtering to reach that same ease-of-use as the rest of our stack. 

We've talked at length about the syntax of these filters - while we've come up with a few options, your :eyes: may see things we haven't. 

## What's your experience?

1. What's your favorite filtering approach? We've looked at many other implementations, including:
- [Firestore](https://cloud.google.com/firestore/docs/query-data/queries)
- [MongoDB](https://docs.mongodb.com/realm/mongodb/query-filters/)
- [PouchDB](https://pouchdb.com/guides/mango-queries.html)
- [IBM Cloudant](https://cloud.ibm.com/docs/Cloudant?topic=Cloudant-query#selector-syntax)

2. What's your biggest headache when filtering documents?

3. Which of our approaches looks the best?

4. What have we missed?  

## Our Approaches

To better demonstrate our filtering approaches, assume we have the following data structure:

```
type Movie = {
    title: string;
    year: number;
    published: boolean;
    director: Director;
    actors: Actor[];
}

type Director = {
    name: string;
}

type Actor = {
    name: string;
}
```

With this schema in mind, we've used three approaches to fetch data in different scenarios, from simple to more advanced:

##### Movie with title 'Iron Man'

```
    db.listDocuments($movieCollection, new Filter()
        .equal('name', 'Iron Man')
    )
    
 
 
    db.listDocuments($movieCollection, { 
      "title: { "equals": "Iron Man" } 
    })
      
  
  
    val filter = new Filter()
          .equal('name', 'Iron Man')
    db.listDocuments($movieCollection, filter)*
```    
* Only so many ways to write simple filters. 🙂 
##### Movies published between 1999 and 2001

```
    db.listDocuments($movieCollection, new Filter()
        .equal('published', true)
        .greaterEquals('year',1999).lesser('year',2001)
    )



    db.listDocuments($movieCollection, {
        "year": {
            "greaterEquals": 1999,
            "lesser": 2001
        }, 
        "published": {
            "equals": true
        }
    })



    val filter = new Filter()
        .equal('published', true)
        .between('year',  include(1999), exclude(2001)) // year >= 1999 && year <2001
    db.listDocuments($movieCollection, filter)
```

##### Movies that actors "Brad Pitt" and "Angelina Jolie" performed together, published after 2000, not Directed by "Michael Bay"

```
    db.listDocuments($movieCollection, new Filter()
        .containsAll("actors.name", ["Brad Pitt",  "Angelina Jolie"])
        .equal("published",true).greater("year",2000)
        .notEqual("director.name","Michael Bay")
    )



    db.listDocuments($movieCollection, {
        "actors":  {
            "name":{
                "containsAll": ["Brad Pitt", "Angelina Jolie"]
            }
        },
        "published": {
            "equals": true
        },
        "directors": {
            "equalsNot": "Michael Bay"
        },
        "year": {
            "greater": 2000
        }
    });



    val filter = new Filter().equal('actors.name', and('Brad Pitt', 'Angelina Jolie'))
        .greaterThan('published', include('2000')) 
        .notEqual('director.name', 'Michael Bay')
    db.listDocuments($movieCollection, filter)
```

**Comment by eldadfux** [comment]
Again and as discussed earlier this week, for me the first approach is by far the most intuitive, desired, and the one that matches our simplicity first approach.

- I think the syntax is clean and short
- I think its more intuitive to have the query driven syntax rather than attribute driven syntax - because we set filters and not necessarily think of a query as something grouped by attributes, even if that's the end result its not how we would think of it in most cases.
- The include/exclude method is not very clear to me and also is part of the global scope which is a very bad practice
- The JSON based syntax is really problematic (or at least add unwanted complexity) in the context of typed script languages, and is not friendly. This is also going to be a problem to match our desire to have consistent syntax across SDK as much as possible.
- The JSON object syntax also looks much longer

**Reply by TorstenDittmann** [reply]
Agree to all of this. Especially when I was researching on how we can provide proper typings.

Soon we are going to introduce Response Models, which in most of the language will make users take advantage of Types/Classes in combination with our SDKs. If we want to offer type safe filter - which is in fact a missing feature with most of backends SDKs - we already have to to ditch the JSON idea, since this can very likely only be accomplished with TypeScript and Flutter. Even though these are probably our most used languages, this would be a downside for the rest of the typed languages.

One approach can be something like this:
```typescript
new Filter<Movie>()
    .equals("name", "Iron Man")
    .greater("year", 2000)
    .notEqual("director.name", "Michael Bay");
```

What might get tricky is making the dot notation type safe, but since this has been already done for many MongoDB clients we should have good resources & examples.

This way we can also ensure same usage across all langauges 👍🏻 

Another thing that came to my mind, what if the filters are getting really complex?

We have 3 options:
- Blaming the data structure if complex queries are required
- Chain like 20+ methods 
- Offer an alternative way to methods, where a more query like object can be passed

Passing this, can be an object - that represents how the filters are saved inside the Filter Class and is passed to the constructor.

This point is nothing urgent, but maybe something to consider in the future 👍🏻

**Reply by TorstenDittmann** [reply]
Btw, we should also consider this Filter to be usable with something that is not a Collection.

We should have consistent filters across Storage, Users, co.

**Reply by christyjacob4** [reply]
@eldadfux  I would not be too concerned about global functions because they can be imported and scoped specifically to the file in which it is being used . Consider this

```ts
// Typescript / JS
import { includes } from 'appwrite/filters'
import { includes as Something } from 'appwrite/filters'
```

```java
// Java or Kotlin
import appwrite.filters.includes;
```

**Reply by eldadfux** [reply]
@christyjacob4 agree, but if we can avoid adding multiple statements for a simple filter I would go that way 😉

**Reply by christyjacob4** [reply]
@eldadfux  Are you talking about reducing the number of imports?

**Reply by eldadfux** [reply]
Yes, but definitely not my main concern as stated above.
